server/swap/swap.go

```diff
@@ -409,37 +409,113 @@ func (s *Swapper) processBlock(block *blockNotification) {
+			return
+		}
+		takerRedeemComplete = confs >= int64(s.coins[match.makerStatus.swapAsset].SwapConf)
+	}
+	return
+}
+
+func (s *Swapper) makerRedeemStatus(match *matchTracker) (makerRedeemComplete bool) {
+	mStatus, tStatus := match.makerStatus, match.takerStatus
+	if !mStatus.redeemTime.IsZero() {
+		confs, err := mStatus.redemption.Confirmations()
+		if err != nil {
+			log.Errorf("Confirmations failed for maker redemption %v: err",
+				mStatus.redemption.TxID(), err) // Severity?
+			return
+		}
+		makerRedeemComplete = confs >= int64(s.coins[tStatus.swapAsset].SwapConf)
+	}
+	return
+}
+
+// TxMonitored determines whether the transaction for the given user is involved
+// in a DEX-monitored trade. Note that the swap contract tx is considered
+// monitored until the swap is complete, regardless of confirms. This allows
+// change outputs from a dex-monitored swap contract to be used to fund
+// additional swaps prior to FundConf. e.g. OrderRouter may allow coins to fund
+// orders where: (coins.confs >= FundConf) OR TxMonitored(coins.tx).
+func (s *Swapper) TxMonitored(user account.AccountID, asset uint32, txid string) bool {
+	s.matchMtx.RLock()
+	defer s.matchMtx.RUnlock()
+
+	for _, match := range s.matches {
+		// The swap contract of either the maker or taker must correspond to
+		// specified asset to be of interest.
+		switch asset {
+		case match.makerStatus.swapAsset:
+			// Maker's swap transaction is the asset of interest.
+			if user == match.Maker.User() && match.makerStatus.swap.TxID() == txid {
+				// The swap contract tx is considered monitored until the swap
+				// is complete, regardless of confirms.
+				return true
+			}
+
+			// Taker's redemption transaction is the asset of interest.
+			_, takerRedeemDone := s.redeemStatus(match)
+			if !takerRedeemDone && user == match.Taker.User() &&
+				match.takerStatus.redemption.TxID() == txid {
+				return true
+			}
+		case match.takerStatus.swapAsset:
+			// Taker's swap transaction is the asset of interest.
+			if user == match.Taker.User() && match.takerStatus.swap.TxID() == txid {
+				// The swap contract tx is considered monitored until the swap
+				// is complete, regardless of confirms.
+				return true
+			}
+
+			// Maker's redemption transaction is the asset of interest.
+			makerRedeemDone := s.makerRedeemStatus(match)
+			if !makerRedeemDone && user == match.Maker.User() &&
+				match.makerStatus.redemption.TxID() == txid {
+				return true
+			}
+		default:
+			continue
+		}
+	}
+
+	return false
+}
+
 // checkInaction scans the swapStatus structures relevant to the specified
 // asset. If a client is found to have not acted when required, a match may be
 // revoked and a penalty assigned to the user.
 func (s *Swapper) checkInaction(assetID uint32) {
 	oldestAllowed := time.Now().Add(-s.bTimeout).UTC()
 	deletions := make([]string, 0)
-	s.matchMtx.RLock()
-	defer s.matchMtx.RUnlock()
+	s.matchMtx.Lock()
+	defer s.matchMtx.Unlock()
```

Comment:
There is a lot going on in this locked section, also in processBlock, but to keep the read and subsequent possible writes (`delete`) atomic, I'm just using a single full lock.

Reply:
Should be good. There are no RPC calls here anyway, so I expect this check to go quick. The only thing that might be remotely slow is the calls to `Penalize`, since there is a DB call there. `revoke` has only coin cache access and non-blocking requests. Is it worth running `Penalize` as a goroutine? Regardless, write lock is okay.

May be a want to poke here during load testing though.

Reply:
> Is it worth running `Penalize` as a goroutine?

It may be, but it's not clear yet that it's worth the potential complications that come with that.